src/lazy/encoder/writer.rs

```diff
@@ -27,34 +28,107 @@ use crate::result::IonFailure;
 use crate::write_config::WriteConfig;
 use crate::{
     AnyEncoding, ContextWriter, Decimal, Element, ElementWriter, Int, IonInput, IonResult, IonType,
-    IonVersion, MacroTable, RawSymbolRef, Reader, Symbol, SymbolTable, TemplateCompiler, Timestamp,
-    UInt, Value,
+    IonVersion, MacroDef, MacroTable, RawSymbolRef, Reader, Symbol, SymbolId, SymbolTable,
+    TemplateCompiler, TemplateMacro, Timestamp, UInt, Value,
 };
 
-pub(crate) struct WriterContext {
```

Comment:
🪧 The symbol table can be modified by the user at any depth by writing a new symbol, annotation, or field name. This required value writer types to pass around a `&mut WriterContext` to make updates as needed.

In contrast, the macro table can only be modified at the top level, between values. When the `ApplicationEExpWriter` goes to store a shared reference to the invoked macro, it causes a conflict. There can't simultaneously be a `&mut WriterContext` _and_ a `&MacroDef` that points inside the `WriterContext`. Therefore, I've split the `WriterContext` into a `WriterSymbolTable` and `WriterMacroTable` that can be managed separately. Each new type is a thin wrapper around `SymbolTable`/`MacroTable` and does some minor bookkeeping each time the underlying table is modified.